### ISSUE TITLE
TypeError: RPC Method is not JSON serializable

### ISSUE THREAD

**Issue by shamanis** [alorence/django-modern-rpc#5]
```
Unhandled exception in thread started by <function check_errors.<locals>.wrapper at 0x7f5c85273510>
Traceback (most recent call last):
  
File "/home/shaman/Work/envs/smart_house/lib/python3.4/site-packages/django/utils/autoreload.py", line 226, in wrapper
    fn(*args, **kwargs)
  
File "/home/shaman/Work/envs/smart_house/lib/python3.4/site-packages/django/core/management/commands/runserver.py", line 109, in inner_run
    autoreload.raise_last_exception()
  
File "/home/shaman/Work/envs/smart_house/lib/python3.4/site-packages/django/utils/autoreload.py", line 249, in raise_last_exception
    six.reraise(*_exception)
  
File "/home/shaman/Work/envs/smart_house/lib/python3.4/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  
File "/home/shaman/Work/envs/smart_house/lib/python3.4/site-packages/django/utils/autoreload.py", line 226, in wrapper
    fn(*args, **kwargs)
  
File "/home/shaman/Work/envs/smart_house/lib/python3.4/site-packages/django/__init__.py", line 18, in setup
    apps.populate(settings.INSTALLED_APPS)
  
File "/home/shaman/Work/envs/smart_house/lib/python3.4/site-packages/django/apps/registry.py", line 115, in populate
    app_config.ready()
  
File "/home/shaman/Work/envs/smart_house/lib/python3.4/site-packages/modernrpc/apps.py", line 43, in ready
    registered_name = register_rpc_method(func)
  
File "/home/shaman/Work/envs/smart_house/lib/python3.4/site-packages/modernrpc/core.py", line 358, in register_rpc_method
    cache.set(RPC_REGISTRY_KEY, registry, timeout=DEFAULT_REGISTRY_TIMEOUT)
  
File "/home/shaman/Work/envs/smart_house/lib/python3.4/site-packages/django_redis/cache.py", line 33, in _decorator
    return method(self, *args, **kwargs)
  
File "/home/shaman/Work/envs/smart_house/lib/python3.4/site-packages/django_redis/cache.py", line 68, in set
    return self.client.set(*args, **kwargs)
  
File "/home/shaman/Work/envs/smart_house/lib/python3.4/site-packages/django_redis/client/default.py", line 109, in set
    nvalue = self.encode(value)
  
File "/home/shaman/Work/envs/smart_house/lib/python3.4/site-packages/django_redis/client/default.py", line 329, in encode
    value = self._serializer.dumps(value)
  
File "/home/shaman/Work/envs/smart_house/lib/python3.4/site-packages/django_redis/serializers/json.py", line 14, in dumps
    return force_bytes(json.dumps(value))
  
File "/usr/lib/python3.4/json/__init__.py", line 230, in dumps
    return _default_encoder.encode(obj)
  
File "/usr/lib/python3.4/json/encoder.py", line 192, in encode
    chunks = self.iterencode(o, _one_shot=True)
  
File "/usr/lib/python3.4/json/encoder.py", line 250, in iterencode
    return _iterencode(o, 0)
  
File "/usr/lib/python3.4/json/encoder.py", line 173, in default
    raise TypeError(repr(o) + " is not JSON serializable")

TypeError: RPC Method add is not JSON serializable
```

Django: 1.9.10
Python: 3.4.2

Example from "Quick start" guide not work.

**Comment by shamanis** [comment]
The issue is resolved.
This problem in django-redis library.
Settings CACHES['default']['OPTIONS']['SERIALIZER'] = 'django_redis.serializers.json.JSONSerializer' causes this error.

**Comment by alorence** [comment]
Hi,

Thank you very much for the feedback.

As you noticed, this app uses the cache configured in Django settings to store the registry of rpc methods. This process is done in `modernrpc.apps.ModernRpcConfig.ready()` method when project starts, so the list of available method remains in cache as long as the app is running. I was able to reproduce the reported error with Python 3.6 and django-redis 4.8.0.

At the moment, RPCMethod instances are designed to be pickled. As a result, the default redis serializer (django_redis.serializers.pickle.PickleSerializer) can successfully store it. But as you noticed, the JSONSerializer (and some others, like MSGPackSerializer) fail to dump the structure.

I keep this ticket open, since in my opinion, the registry should be able to be serialized in all case, whatever cache settings is defined in a project. I will work on this ASAP, and will try to add tests for the most used cache systems (memcached, redis, DB cache, etc). This should also be mentionned in the documentation.

In the meantime, you can fall to the default PickleSerializer or use the defaut django cache backend (in memory) to use the library. A new version fixing those limitations should be available in the next days (or weeks)

**Comment by alorence** [comment]
After some performance tests, I realized that the use of cache system to store the registry of decorated RPC methods was useless. In addition, this system increased the time needed to get a method from the registry.
Use of cache system was causing more issue than it solved, so I decided to remove it.

A new version 0.6.0 has just been pushed on Pypi. It should completely solve the reported issue, and will have better performances. Let me know if you find anything problematic with this new version.

Best regards
